File: PR Kelas/18 mar/KoleksiKendaraan.cpp
#include "KoleksiKendaraan.hpp"
#include <iostream>
using namespace std;

KoleksiKendaraan::KoleksiKendaraan(){
    this->size=100;
    this->neff=0;
    this->arr = new Kendaraan[this->size];
}

KoleksiKendaraan::KoleksiKendaraan(int size){
    this->size=size;
    this->neff=0;
    this->arr = new Kendaraan[this->size];
}

KoleksiKendaraan::KoleksiKendaraan(const KoleksiKendaraan &k){
    this->size=k.size;
    this->neff=k.neff;
    this->arr = new Kendaraan[this->size];
    for (int i=0; i<k.neff; i++) {
        this->arr[i]=k.arr[i];
    }
}

KoleksiKendaraan& KoleksiKendaraan::operator=(const KoleksiKendaraan& k){
    if (this != &k) {
        delete[] this->arr;
        this->size=k.size;
        this->neff=k.neff;
        this->arr = new Kendaraan[this->size];
        for (int i=0; i<k.neff; i++) {
            this->arr[i]=k.arr[i];
        }
    }
    return *this;
}

KoleksiKendaraan::~KoleksiKendaraan() {
    delete[] this->arr;
}

void KoleksiKendaraan::printAll() {
    for (int i=0; i<this->neff; i++) {
        this->arr[i].printInfo();
    }
}

KoleksiKendaraan& KoleksiKendaraan::operator<<(const Kendaraan &k) {
    this->arr[this->neff]=k;
    this->neff++;
    return *this;
}

KoleksiKendaraan& KoleksiKendaraan::operator<<(const KoleksiKendaraan &k) {
    for (int i=0; i<k.neff; i++) {
        if (this->neff < this->size) {
            this->arr[this->neff]=k.arr[i];
            this->neff++;
        }
    }
    return *this;
}

File: PR Kelas/13 mei/KoleksiKendaraan.java
import java.util.ArrayList;
import java.util.List;

class KoleksiKendaraan {
    private int size;
    private int neff;
    private List<Kendaraan> arr;
    public KoleksiKendaraan() {
        this.size=100;
        this.neff=0;
        this.arr=new ArrayList<>(size);
    }
    public KoleksiKendaraan(int s) {
        this.size=s;
        this.neff=0;
        this.arr=new ArrayList<>(size);
    }
    public KoleksiKendaraan(KoleksiKendaraan k) {
        this.size = k.size;
        this.neff = k.neff;
        this.arr = new ArrayList<>();
        for (Kendaraan kendaraan : k.arr) {
            this.arr.add(new Kendaraan(kendaraan));
        }
    }
    public void copyFrom(KoleksiKendaraan k) {
        this.size = k.size;
        this.neff = k.neff;
        this.arr = new ArrayList<>();
        for (Kendaraan kendaraan : k.arr) {
            this.arr.add(new Kendaraan(kendaraan));
        }
    }
    // Tidak ada destruktor
    public void printAll() {
        for (Kendaraan k : arr) {
            k.printInfo();
            System.out.println();
        }
    }
    public void add(Kendaraan k) {
        if (neff < size) {
            arr.add(new Kendaraan(k));
            neff++;
        } else {
            System.out.println("Koleksi penuh. Tidak bisa menambahkan kendaraan lagi.");
        }
    }
    public void addAll(KoleksiKendaraan k) {
        for (Kendaraan kendaraan : k.arr) {
            if (neff < size) {
                arr.add(new Kendaraan(kendaraan));
                neff++;
            } else {
                System.out.println("Koleksi penuh. Tidak semua kendaraan berhasil ditambahkan.");
                break;
            }
        }
    }
}
